### DISCUSSION TITLE
"Petzold" for WinUI 3?

### DISCUSSION THREAD

**Discussion by meirumeiru** [microsoft/WindowsAppSDK#2102] [Q&A]
Is there something like the "Petzold" for WinUI 3? Where the message handling mechanism is explained and the loops and everything? Or did Raymond Chen already write about it?

I've problems with understanding how messages are handled. How many thread do we have? What about inter-thread communication? What if I want to create an object-class just to run in-sync with the UI (in Win32 I created an invisible window class for that)? What if I don't want async for some operations (yes, that's true, while all my applications do have a lot of threads for background tasks, sometimes I do things inside my main GUI thread because I want it to block! ... for some micro-seconds).

The documentation I find is often not going deep enough. And most of the time I don't know what's true now. Is XAML / XAML islands of WPF compareable to WinUI 3? Or not? Why is this feature still experimental? Where's the problem? I don't really see it (because I don't understand the meachnism behind it... of course... but... I'd like to know everything).

Would be great to get some good documentation... or books if the exist. :-) ... or an explanation about the message handling mechanism (e.g. what happens, when you click on a window... does it start with a WM_LBUTTONDOWN? and WinUI, XAML handles this or ... isn't there a message involved at all?)

**Comment by castorix** [comment]
In WinUI, _Microsoft::UI::Dispatching::DispatcherQueue_ can be used, as explained at [Threading functionality migration](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/guides/threading)
WinUI XAML controls  ([Microsoft.UI.Xaml.Controls](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls?view=winui-3.0)) are about the same as UWP XAML controls ([Windows.UI.Xaml.Controls](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls?view=winrt-22000))

**Comment by FinnurHrafn** [comment]
I am just beginning my WinUI 3 journey. I have found this book very helpful:
https://www.amazon.com/Learn-WinUI-3-0-application-development/dp/1800208669
I am not sure it answers all your questions but many topics are covered well, mostly relevant even though it almost 1 year old.

**Comment by ghost** [comment]
June 9 is the release date. 

### "Modernizing your Windows Applications with Windows Apps SDK and WinUI 3"
Author : MSFT's Matteo Pagani
https://www.amazon.com/Modernizing-your-Windows-Applications-WinUI-ebook/dp/B09R4RJFP4

**Reply by mgradwohl** [reply]
Wish Matteo's book was C++/WinRT
